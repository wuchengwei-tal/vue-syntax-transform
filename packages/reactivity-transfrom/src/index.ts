// API
export { parse } from "./parse";
export { compileTemplate } from "./compileTemplate";
export { compileScript } from "./compileScript";
export { rewriteDefault } from "./rewriteDefault";
export {
    shouldTransform as shouldTransformRef,
    transform as transformRef,
    transformAST as transformRefAST,
} from "@vue/reactivity-transform";

// Utilities
export { parse as babelParse } from "@babel/parser";
import MagicString from "magic-string";
export { MagicString };
// technically internal but we want it in @vue/repl, cast it as any to avoid
// relying on estree types
import { walk as _walk } from "estree-walker";
export const walk = _walk as any;

// Types
export {
    SFCParseOptions,
    SFCParseResult,
    SFCDescriptor,
    SFCBlock,
    SFCTemplateBlock,
    SFCScriptBlock,
    SFCStyleBlock,
} from "./parse";
export {
    TemplateCompiler,
    SFCTemplateCompileOptions,
    SFCTemplateCompileResults,
} from "./compileTemplate";

export { SFCScriptCompileOptions } from "./compileScript";
export {
    AssetURLOptions,
    AssetURLTagConfig,
} from "./templateTransformAssetUrl";
