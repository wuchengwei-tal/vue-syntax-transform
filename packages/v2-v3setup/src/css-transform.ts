import type { CssNode, Dimension, Raw, Selector } from 'css-tree'
import MagicString from 'magic-string'
const csstree = require('css-tree')

const Ratio = 100

export function cssTransform(css: string) {
  return _cssTransform(transComment(css))
}

export function _cssTransform(css: string, raw = '') {
  const s = new MagicString(css)
  const ast = csstree.parse(css, { positions: true })

  function transDimension(node: Dimension) {
    if (node.unit === 'rem') {
      let { start, end } = node.loc!
      const val = parseFloat(node.value) * Ratio + 'px'
      s.overwrite(start.offset, end.offset, val)
    }
  }

  function transSelector(node: Selector) {
    const head = node.children.first
    if (head?.type === 'Combinator' && /\/\s*deep\s*\//.test(head?.name)) {
      const { start, end } = head.loc!
      s.overwrite(start.offset, end.offset, ':deep(')
      s.appendRight(node.loc!.end.offset, ')')
    }
  }

  function transRaw(node: Raw) {
    if (node.value !== raw) {
      const { start, end } = node.loc!
      const css = _cssTransform(node.value, node.value)
      s.overwrite(start.offset, end.offset, css)
    }
  }

  csstree.walk(ast, {
    leave(node: CssNode) {
      if (node.type === 'Dimension') transDimension(node)

      if (node.type === 'Selector') transSelector(node)

      if (node.type === 'Raw') transRaw(node)
    }
  })

  return s.toString()
}

export function transComment(css: string) {
  let code = ''
  for (let line of css.split('\n')) {
    if (/\s*\/\/.*/.test(line)) {
      line = line.replace('//', '/*')
      line += ' */'
    }

    code += line + '\n'
  }
  return code
}
