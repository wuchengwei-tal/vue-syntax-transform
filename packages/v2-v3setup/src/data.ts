export enum BindingTypes {
  /**
   * returned from data()
   */
  DATA = 'data',
  /**
   * declared as a prop
   */
  PROPS = 'props',
  /**
   * a let binding
   */
  LET = 'let',
  /**
   * a const binding that.
   */
  CONST = 'const',

  COMPUTED = 'computed',

  FILTER = 'filter',

  METHOD = 'method',

  WATCH = 'watch',

  HOOK = 'hook',

  $ = '$'
}

export type BindingMap<T> = Record<string, { type: BindingTypes; value: T }>

export const LifeCircleHookMap = {
  beforeMount: 'onBeforeMount',
  mounted: 'onMounted',
  beforeUpdate: 'onBeforeUpdate',
  updated: 'onUpdated',
  activated: 'onActivated',
  deactivated: 'onDeactivated',
  beforeDestroy: 'onBeforeUnmount',
  destroyed: 'onUnmounted',
  beforeRouteEnter: 'onBeforeRouteUpdate',
  beforeRouteLeave: 'onBeforeRouteLeave',
  beforeCreate: '',
  created: ''
}

export const WalkOrder = [
  'data',
  'methods',
  'computed',
  'filter',
  'watch',
  ...Object.keys(LifeCircleHookMap)
]

export const OutputOrder = [
  BindingTypes.CONST,
  BindingTypes.LET,
  BindingTypes.PROPS,
  BindingTypes.DATA,
  BindingTypes.COMPUTED,
  BindingTypes.FILTER,
  BindingTypes.METHOD
]

export const RenderFunction = 'RenderFunction'
