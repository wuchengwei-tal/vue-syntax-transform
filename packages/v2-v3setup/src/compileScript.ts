import { parse as _parse, ParserOptions } from '@babel/parser'

import {
  Node,
  ObjectExpression,
  ObjectMethod,
  ObjectProperty,
  Program,
  BlockStatement,
  Identifier,
  StringLiteral,
  Expression,
  Declaration,
  ArrayPattern,
  ObjectPattern,
  identifier
} from '@babel/types'

import MagicString from 'magic-string'

import { generateCodeFrame } from '@vue/shared'

import { BindingTypes } from '@vue-transform/shared'
import { BindingMap, RenderFunction, VModel } from './data'
import { transformBindings, registerBinding } from './transform'

export function compileScript(
  source: string,
  options: { id: string }
): { content: string; bindings?: Record<string, BindingTypes> } {
  if (!source) return { content: source }

  // metadata that needs to be returned
  const bindingMetadata: Record<string, BindingTypes> = {}
  const optionsBindings: BindingMap<any> = Object.create(null)
  const watcherBindings: BindingMap<ObjectProperty['value'] | ObjectMethod> =
    Object.create(null)
  const hookBindings: BindingMap<ObjectMethod> = Object.create(null)
  const model: VModel = { exist: false, prop: 'value', event: 'input' }

  let name = ''
  const mixins = []
  const assets: string[] = []
  const components: { alias?: string; name: string }[] = []
  let defaultExport: Node | undefined
  const s = new MagicString(source)
  const startOffset = 0
  const endOffset = source.length

  function parse(
    input: string,
    _options: ParserOptions,
    offset: number
  ): Program {
    try {
      return _parse(input, _options).program
    } catch (e: any) {
      e.message = `[@vue/compiler-sfc] ${e.message}\n\n${
        options.id
      }\n${generateCodeFrame(source, e.pos + offset, e.pos + offset + 1)}`
      throw e
    }
  }

  function walkOptions(node: ObjectExpression) {
    for (const property of node.properties) {
      if (property.type !== 'SpreadElement') {
        walkOption(property)
      } else {
        if (property.argument.type === 'ObjectExpression')
          walkOptions(property.argument)
      }
    }
  }

  function walkOption(property: ObjectProperty | ObjectMethod) {
    const { key } = property
    let name = ''
    if (key.type === 'Identifier') name = key.name
    if (key.type === 'StringLiteral') name = key.value
    if (property.type === 'ObjectMethod') {
      if (name === 'data') walkState(optionsBindings, property.body, assets)
      else if (name === 'render')
        registerBinding(
          optionsBindings,
          identifier(RenderFunction),
          property,
          BindingTypes.METHOD
        )
      else {
        registerBinding(
          hookBindings,
          identifier(name),
          property,
          BindingTypes.HOOK
        )
      }
    }

    if (property.type === 'ObjectProperty') {
      if (name === 'name' && property.value.type === 'StringLiteral')
        name = property.value.value

      if (name === 'components') {
        const { value } = property
        if (value.type === 'ObjectExpression') {
          for (const component of value.properties) {
            if (component.type === 'ObjectProperty') {
              const { key, value } = component
              if (key.type === 'Identifier' && value.type === 'Identifier') {
                const { name } = value
                const alias = key.name
                alias === name
                  ? components.push({ name })
                  : components.push({ alias, name })
              }
            }
          }
        }
      }

      if (name === 'mixins') {
        const { value } = property
        if (value.type === 'ArrayExpression') {
          for (const mixin of value.elements) {
            if (mixin?.type === 'Identifier') mixins.push(mixin.name)
          }
        }
      }

      if (name === 'props') walkProps(optionsBindings, property)

      if (name === 'model') updateModel(model, property)

      if (name === 'filters')
        walkMethods(optionsBindings, property, BindingTypes.FILTER)

      if (name === 'computed') walkGetter(optionsBindings, property)

      if (name === 'methods')
        walkMethods(optionsBindings, property, BindingTypes.METHOD)

      if (name === 'watch') walkWatches(watcherBindings, property)
    }
  }

  function walkProps(bindings: BindingMap<any>, property: ObjectProperty) {
    const { value } = property
    if (value.type === 'ObjectExpression') {
      if (value.properties.length) {
        const props = property.key as Identifier
        registerBinding(bindings, props, value, BindingTypes.$)
      }
      for (const p of value.properties) {
        if (p.type === 'ObjectProperty') {
          const { key, value } = p
          if (key.type === 'Identifier') {
            if (key.name === model.prop) {
              if (!model.exist) {
                model.exist = /emit\(('|")input/.test(source)
              }
              registerBinding(
                bindings,
                identifier(key.name),
                value,
                BindingTypes.PROPS
              )
              key.name = 'modelValue'
            } else registerBinding(bindings, key, null, BindingTypes.PROPS)
          }
        }
      }
    }
  }

  // 0. parse both <script> blocks
  const scriptAst = parse(
    source,
    { plugins: [], sourceType: 'module' },
    startOffset
  )
  if (!scriptAst) return { content: source }

  for (const node of scriptAst.body) {
    if (node.type === 'ImportDeclaration') {
      const { source, specifiers } = node
      for (const specifier of specifiers) {
        if (specifier.type === 'ImportDefaultSpecifier') {
          if (!/\.(vue|js)/.test(source.value) && source.value.includes('.'))
            assets.push(specifier.local.name)
        }
      }
    }

    if (node.type === 'ExportNamedDeclaration') {
      s.remove(startOffset + node.start!, startOffset + node.end!)
    }

    if (node.type === 'VariableDeclaration') {
      walkDeclaration(node, optionsBindings)
    }

    if (node.type === 'ExportDefaultDeclaration') {
      const { declaration } = node

      if (declaration.type === 'ObjectExpression') {
        walkOptions(declaration)
      }

      // export default
      defaultExport = node
    }
  }

  // . remove non-script content
  s.remove(0, startOffset)
  s.remove(endOffset, source.length)

  for (const key in optionsBindings) {
    bindingMetadata[key] = optionsBindings[key].type
  }

  const code = transformBindings(
    optionsBindings,
    watcherBindings,
    hookBindings,
    model
  )

  if (defaultExport) {
    s.prependRight(endOffset, code + '\n')
    s.remove(
      startOffset + defaultExport.start!,
      startOffset + defaultExport.end!
    )
  }

  s.trim()

  return {
    bindings: bindingMetadata,
    content: s.toString()
  }
}

function walkState(
  bindings: BindingMap<ObjectProperty['value']>,
  block: BlockStatement,
  assets: string[]
) {
  for (const node of block.body) {
    if (node.type === 'ReturnStatement') {
      const { argument } = node
      if (argument?.type === 'ObjectExpression') {
        for (const state of argument.properties) {
          if (state.type === 'ObjectProperty') {
            const { key, value } = state
            if (key.type === 'Identifier' && !assets.includes(key.name))
              registerBinding(bindings, key, value, BindingTypes.DATA)
          }
        }
      }
    }
  }
}

function walkGetter(
  bindings: BindingMap<BlockStatement | string | Expression>,
  computed: ObjectProperty
) {
  if (computed.value.type === 'ObjectExpression') {
    for (const property of computed.value.properties) {
      if (property.type === 'ObjectMethod') {
        const { key, body } = property
        if (key.type === 'Identifier')
          registerBinding(bindings, key, body, BindingTypes.COMPUTED)
      }

      if (property.type === 'ObjectProperty') {
        const { key, value } = property
        if (key.type === 'Identifier') {
          if (
            value.type === 'FunctionExpression' ||
            value.type === 'ArrowFunctionExpression'
          ) {
            registerBinding(bindings, key, value.body, BindingTypes.COMPUTED)
          }
        }
      }

      if (property.type === 'SpreadElement') {
        const { argument } = property
        if (argument.type === 'CallExpression') {
          const { callee } = argument
          if (callee.type === 'Identifier' && callee.name === 'mapGetters') {
            const items = argument.arguments
            if (items.length && items[0].type === 'ObjectExpression') {
              for (const item of items[0].properties) {
                if (item.type !== 'ObjectProperty') continue
                const { key, value } = item
                const val = (value as StringLiteral).value
                if (key.type === 'Identifier')
                  registerBinding(bindings, key, val, BindingTypes.COMPUTED)
              }
            }
          }
        }
      }
    }
  }
}

function walkMethods(
  bindings: BindingMap<ObjectMethod>,
  property: ObjectProperty,
  type: BindingTypes
) {
  const { value } = property
  if (value.type === 'ObjectExpression') {
    for (const func of value.properties) {
      if (func.type === 'ObjectMethod') {
        if (func.key.type === 'Identifier')
          registerBinding(bindings, func.key, func, type)
      }
    }
  }
}

function walkWatches(
  bindings: BindingMap<ObjectProperty['value'] | ObjectMethod>,
  property: ObjectProperty
) {
  const { value } = property
  if (value.type === 'ObjectExpression') {
    for (const watcher of value.properties) {
      if (watcher.type === 'ObjectProperty') {
        const { key, value } = watcher
        if (key.type === 'Identifier')
          registerBinding(bindings, key, value, BindingTypes.WATCH)

        if (key.type === 'StringLiteral') {
          registerBinding(
            bindings,
            identifier(key.value),
            value,
            BindingTypes.WATCH
          )
        }
      }
      if (watcher.type === 'ObjectMethod') {
        if (watcher.key.type === 'Identifier')
          registerBinding(bindings, watcher.key, watcher, BindingTypes.WATCH)
      }
    }
  }
}

function walkDeclaration(node: Declaration, bindings: BindingMap<any>) {
  if (node.type === 'VariableDeclaration') {
    const isConst = node.kind === 'const'
    const bindingType = isConst ? BindingTypes.CONST : BindingTypes.LET
    for (const { id, init } of node.declarations) {
      if (id.type === 'Identifier') {
        registerBinding(bindings, id, init, bindingType)
      }
      if (id.type === 'ObjectPattern') {
        walkObjectPattern(id, bindings, isConst)
      } else if (id.type === 'ArrayPattern') {
        walkArrayPattern(id, bindings, isConst)
      }
    }
  } else if (
    node.type === 'FunctionDeclaration' ||
    node.type === 'ClassDeclaration'
  ) {
    // export function foo() {} / export class Foo {}
    // export declarations must be named.
    registerBinding(bindings, node.id!, node.body, BindingTypes.CONST)
  }
}

function walkObjectPattern(
  node: ObjectPattern,
  bindings: BindingMap<any>,
  isConst: boolean
) {
  for (const p of node.properties) {
    if (p.type === 'ObjectProperty') {
      if (p.key.type === 'Identifier' && p.key === p.value) {
        // shorthand: const { x } = ...
        const type = isConst ? BindingTypes.CONST : BindingTypes.LET
        registerBinding(bindings, p.key, p.value, type)
      } else {
        walkPattern(p.value, bindings, isConst)
      }
    } else {
      // ...rest
      // argument can only be identifier when destructuring
      const type = isConst ? BindingTypes.CONST : BindingTypes.LET
      registerBinding(bindings, p.argument as Identifier, '', type)
    }
  }
}

function walkArrayPattern(
  node: ArrayPattern,
  bindings: BindingMap<any>,
  isConst: boolean
) {
  for (const e of node.elements) {
    e && walkPattern(e, bindings, isConst)
  }
}

function walkPattern(node: Node, bindings: BindingMap<any>, isConst: boolean) {
  if (node.type === 'Identifier') {
    const type = isConst ? BindingTypes.CONST : BindingTypes.LET
    registerBinding(bindings, node, '', type)
  } else if (node.type === 'RestElement') {
    // argument can only be identifier when destructuring
    const type = isConst ? BindingTypes.CONST : BindingTypes.LET
    registerBinding(bindings, node.argument as Identifier, '', type)
  } else if (node.type === 'ObjectPattern') {
    walkObjectPattern(node, bindings, isConst)
  } else if (node.type === 'ArrayPattern') {
    walkArrayPattern(node, bindings, isConst)
  } else if (node.type === 'AssignmentPattern') {
    if (node.left.type === 'Identifier') {
      const type = isConst ? BindingTypes.CONST : BindingTypes.LET
      registerBinding(bindings, node.left, '', type)
    } else {
      walkPattern(node.left, bindings, isConst)
    }
  }
}

function updateModel(model: VModel, property: ObjectProperty) {
  model.exist = true
  const { value } = property
  if (value.type === 'ObjectExpression') {
    for (const p of value.properties) {
      if (p.type === 'ObjectProperty') {
        const { key, value } = p
        if (key.type === 'Identifier' && value.type === 'StringLiteral') {
          if (key.name === 'prop') model.prop = value.value
          if (key.name === 'event') model.event = value.value
        }
      }
    }
  }
}
