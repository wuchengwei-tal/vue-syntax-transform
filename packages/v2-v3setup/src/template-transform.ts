import MagicString from 'magic-string'
import {
  type ASTElement,
  type SFCBlock,
  parseComponent,
  compile,
  ASTNode
} from 'vue-template-compiler'
import { Comment } from './data'

export function templateTransform(source: string) {
  const descriptor = parseComponent(source)
  const { template } = descriptor
  if (!template) return ''

  let inheritAttrs = true
  if (descriptor.script) {
    const { content } = descriptor.script
    const match = content.match(/inheritAttrs\s*:\s*(true|false)/)
    if (match) inheritAttrs = match[1] === 'true'
  }

  const { ast } = compile(template.content.trim(), { outputSourceRange: true })
  return transform({ ast, ...template, inheritAttrs })
}

type Template = { ast?: ASTElement; inheritAttrs: boolean } & SFCBlock

function transform(template: Template) {
  if (!template || !template.ast) return ''
  const { ast, content, inheritAttrs } = template
  const s = new MagicString(content)

  const offset = 1
  // console.log(ast)

  function walk(ast: ASTElement) {
    transNode(ast)
    ast.children.forEach(node => {
      if (node.type === ELEMENT) {
        walk(node)
      } else if (node.type === TEXT) {
      } else if (node.type === Expression) {
      }
    })
  }

  function transNode(node: ASTElement) {
    node.attrs?.forEach((attr, i) => {
      // .sync -> v-model
      if (node.attrsList[i]?.name === `:${attr.name}.sync`) {
        const vModel = `v-model:${attr.name}="${attr.value}"`
        // @ts-ignore
        s.overwrite(attr.start + offset, attr.end + offset, vModel)
      }
    })

    const vBind = node.attrsList.find(attr => attr.name === 'v-bind')
    if (vBind || node.nativeEvents) {
      const events = Object.keys(node?.nativeEvents || {})

      node.attrsList.forEach(attr => {
        // @ts-ignore
        const { start, end, name } = attr

        // console.log(node)
        // rm v-on.native
        events.forEach(event => {
          if (name.includes(`v-on:${event}`) || name.includes(`@${event}`)) {
            let value = s.original.slice(start + offset, end + offset)
            value = value.replace(`.native`, '')
            // console.log(value, start, end)
            s.overwrite(start + offset, end + offset, value)
          }
        })

        // v-bind order
        // @ts-ignore
        if (start < vBind?.start) {
          const value = s.toString().slice(start + offset, end + offset)
          s.remove(start + offset, end + offset)
          // @ts-ignore
          s.appendRight(vBind.end + offset, ' ' + value)
        }
      })
    }

    transDirective(node)

    removeKeyAttr(node)

    transIsAttr(node)
  }

  function removeKeyAttr(node: ASTElement) {
    const { ifConditions } = node

    const getKey = (n: ASTNode) =>
      // @ts-ignore
      n.rawAttrsMap?.['key'] || n.rawAttrsMap?.[':key']

    if (node['for']) {
      // console.log(node.children, node.alias)
      // rm v-for children's key
      node.children.forEach(child => {
        if (child.type !== ELEMENT || !child.key) return
        const key = getKey(child)
        if (key) {
          const { start, end } = key
          s.remove(start + offset, end + offset)
          delete child?.key
        }
      })

      // @ts-ignore
      const vFor = node?.rawAttrsMap?.['v-for']
      if (vFor && !node.key) {
        const { start, end } = vFor
        let { alias, iterator1 } = node
        if (!iterator1) iterator1 = 'i_' + alias

        const value = `v-for="(${alias}, ${iterator1}) in ${node['for']}"`
        s.overwrite(start + offset, end + offset, value)
        s.appendRight(end + offset, ` :key="${iterator1}"`)
      }

      // add Migration Strategy comment for v-if and v-for
      if (ifConditions?.length) {
        // @ts-ignore
        s.appendLeft(node.start, Comment.vForAndVIf)
      }
    }

    // rm v-if's key
    if (ifConditions?.length) {
      ifConditions.forEach(condition => {
        if (condition?.block?.key) {
          // console.log(condition)
          const { block } = condition

          const key = getKey(block)
          if (key) {
            const { start, end } = key
            s.remove(start + offset, end + offset)
          }
        }
      })
    }
  }

  function transDirective(node: ASTElement) {
    const { directives, attrsList } = node
    if (!directives) return

    directives.forEach(directive => {
      // @ts-ignore
      const { name, value, start, end } = directive
      if (name === 'on') {
        if (value === '$listeners') {
          s.remove(start + offset, end + offset)
        }
      }

      if (name === 'bind' && value === '$attrs' && !inheritAttrs) {
        // @ts-ignore
        s.appendLeft(node.start + offset, Comment.inheritAttrsFalse)
      }
    })
  }

  function transIsAttr(node: ASTElement) {
    if (node.attrsMap['is']) {
      // @ts-ignore
      const { tag, start, end } = node
      if (tag !== 'component') {
        let value = s.original.slice(start + offset, end + offset)
        value = value.replace(`<${tag}`, '<component')
        value = value.replace(`</${tag}>`, '</component>')
        s.overwrite(start + offset, end + offset, value)
      }
    }
  }

  walk(ast)
  // console.log(s.toString())
  return s.toString()
}

const ELEMENT = 1
const Expression = 2
const TEXT = 3
