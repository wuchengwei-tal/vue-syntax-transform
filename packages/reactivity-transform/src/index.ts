export { compileScript } from './compileScript'

import { compileScript } from './compileScript'

export { SFCScriptCompileOptions } from './compileScript'

import { parse } from '@vue/compiler-sfc'

export function reactivityTransform(src: string, id = 'xxxxxx') {
  const { descriptor } = parse(src)
  return compileScript(descriptor, {
    id,
    reactivityTransform: true
  })
}
