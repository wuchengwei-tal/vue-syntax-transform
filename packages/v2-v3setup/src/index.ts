import {
  type SFCDescriptor,
  parseComponent,
  compile
} from 'vue-template-compiler'

import { compileScript } from './compileScript'

import { templateTransform } from './template-transform'
import { cssTransform } from './css-transform'
import MagicString from 'magic-string'

export * from './data'

type TransformedSFC = {
  template?: ReturnType<typeof templateTransform>
  script?: ReturnType<typeof compileScript>
  styles?: ReturnType<typeof cssTransform>[]
  content: string
}

export function sfcTransform(sfc: string, id = '') {
  const descriptor = parseComponent(sfc) as SFCDescriptor & {
    source: string
    cssVars: string[]
    errors: any[]
  }

  const { script, template, styles, source } = descriptor

  const s = new MagicString(source)

  const ret: TransformedSFC = { content: '', styles: [] }

  let inheritAttrs = true
  if (script) {
    // console.log(descriptor)
    const match = script.content.match(/inheritAttrs\s*:\s*(true|false)/)
    if (match) inheritAttrs = match[1] === 'true'
  }

  if (template) {
    const { content, start, end } = template

    const { ast } = compile(content, {
      outputSourceRange: true,
      whitespace: 'preserve'
    })
    ret.template = templateTransform({ ast, ...template, inheritAttrs })
    s.overwrite(start!, end!, ret.template.content)
  }

  if (script) {
    const { start, end, content } = script
    if (content) {
      ret.script = compileScript(content, { id })
      s.overwrite(start!, end!, '\n' + ret.script.content + '\n')
      let i = start!
      while (source[i--] === '>');
      while (source[--i] !== '<');
      if (source.slice(i, i + 7) === '<script') {
        s.prependRight(i + 7, ' setup')
      }
    }
  }

  if (styles.length) {
    styles.forEach(style => {
      const { start, end, content } = style
      if (content) {
        const res = cssTransform(content)
        ret.styles?.push(res)
        s.overwrite(start!, end!, res)
      }
    })
  }

  return { ...ret, content: s.toString() }
}

function transformActions(code: string) {
  const fnRe = /(export )?const \w+ = (async )?\(.*[\)|\n]/g

  let match
  let last = 0
  let output = ''
  while ((match = fnRe.exec(code))) {
    const pre = code.slice(last, match.index)

    let [line, _, __] = match

    last = match.index + line.length

    let flag = line[line.length - 2]
    // console.log(line);
    if (line) {
      // oneline params
      if (flag === '{') {
        const match = /\(((\{.*\})|(\w+)),? ?(.*)?\)/.exec(line)
        if (match) {
          let [arg1, _, __, ___, arg2e] = match
          line = line.replace(arg1.slice(1, -1), arg2e || '')
        }
      }
      output += pre + line
      // multi line params
      if (flag === '(') {
        const params = parseInside(code, last, '(')
        let i = 0
        while (params[i++] !== '}');
        if (params[i] === ',') i++
        output += params.slice(i)

        last += params.length
        while (code[last++] !== '{') output += code[last]
        flag = '{'
      }
    }

    if (flag === '{') {
      let body = parseInside(code, last, '{')
      last += body.length
      body = replaceActionBody(body)
      body = replaceState(body)
      output += body
    }
  }
  output += code.slice(last)

  // console.log(output);
  return output
}

const Pair = { '{': '}', '(': ')', '[': ']' }
function parseInside(code: string, s: number, p: keyof typeof Pair) {
  let e = s
  let cnt = 0
  while (e < code.length - 1) {
    const ch = code[e++]
    if (ch === p) cnt++
    if (ch === Pair[p]) {
      if (cnt) cnt--
      else break
    }
  }
  const body = code.slice(s, e)
  return body
}

function replaceActionBody(code: string) {
  const dispatchRe = /(((store\.)?dispatch)|(commit))\('?((\w+\.?\w+))'?,? ?/g

  let match
  let last = 0
  let output = ''
  while ((match = dispatchRe.exec(code))) {
    const pre = code.slice(last, match.index)

    let [line, _, store, dispatch, commit, fn] = match

    last = match.index + line.length

    if (line && fn) {
      if (fn.includes('.')) fn = fn.split('.')[1]
      line = fn + '('
      // console.log(line)
    }

    output += pre + line
  }
  output += code.slice(last)
  // console.log(output);
  return output
}

function transformMutations(code: string) {
  const { body, index } = getExportDefault(code)
  if (!body) return code

  let last = index
  let output = code.slice(0, index)

  let match
  const mutationRe = /((\w+)|\[types.(\w+)\])(\(state,?.*\)) {/g
  while ((match = mutationRe.exec(body))) {
    let pre = ''
    if (last !== index) pre = body.slice(last, match.index)

    let [line, _, name, type, params] = match

    last = match.index + line.length
    params = params.replace(/state,? ?/, '')
    line = 'export function ' + (name || type) + params + '{'

    let mutationBody = parseInside(body, last, '{')
    last += mutationBody.length + 1 // del comma

    mutationBody = replaceState(mutationBody)

    output += pre + line + mutationBody
  }

  return output
}

function getExportDefault(code: string) {
  const defaultRe = /export default {/
  const match = defaultRe.exec(code)
  if (match) {
    return {
      body: parseInside(code, match.index, '{'),
      index: match.index
    }
  }
  return {}
}

function replaceState(code: string) {
  const stateRe = /(state\.\w+)( |\(|\.|\[|,)/g
  let last = 0
  let output = ''
  let match
  while ((match = stateRe.exec(code))) {
    const pre = code.slice(last, match.index)

    let [res, state] = match
    last = match.index + res.length

    if (res.endsWith('.')) res = res.replace(res, `${res}value.`)
    else res = res.replace(state, `${state}.value`)
    res = res.replace('state.', '')

    output += pre + res
  }
  output += code.slice(last)

  output = output.replace(/rootState\??\.\w+\??\./g, '')

  return output
}
