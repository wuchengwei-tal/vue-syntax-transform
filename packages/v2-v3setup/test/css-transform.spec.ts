import { expect } from 'vitest'
import { cssTransform } from '../src/css-transform'

describe('css transform', () => {
  it('should transform deep', () => {
    const res = cssTransform(`
     /deep/ .foo {
         width: 100px;

         /deep/ .bar {
                height: 100px;
         }
     }
    `)

    expect(res).toMatchSnapshot()
  })

  it('should transform px', () => {
    const res = cssTransform(`
.v-enter,
.v-leave-to {
  opacity: 0;
}

.v-leave,
.v-enter-to {
  opacity: 1;
}
`)

    expect(res).toMatchSnapshot()
  })
})
