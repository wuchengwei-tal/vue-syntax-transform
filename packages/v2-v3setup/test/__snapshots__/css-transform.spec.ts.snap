// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`css transform > should transform deep 1`] = `
"
     :deep( .foo) {
         width: 1rem;

         :deep( .bar) {
                height: 1rem;
         }
     }
    
"
`;

exports[`css transform > should transform px 1`] = `
"
v-enter-from,
.v-leave-to {
  opacity: 0;
}

v-leave-from,
.v-enter-to {
  opacity: 1;
}

"
`;
